jaxopt/_src/lbfgs.py

```diff
@@ -0,0 +1,312 @@
+                        gamma: float = 1.0):
+  """Product between an approximate Hessian inverse and a pytree.
+
+  Histories are pytrees of the same structure as `pytree`.
+  Leaves are arrays of shape `(history_size, ...)`, where
+  `...` means the same shape as `pytree`'s leaves.
+
+  The notation follows the reference below.
+
+  Args:
+    pytree: pytree to multiply with.
+    s_history: pytree with the same structure as `pytree`.
+      Leaves contain parameter residuals, i.e., `s[k] = x[k+1] - x[k]`.
+    y_history: pytree with the same structure as `pytree`.
+      Leaves contain gradient residuals, i.e., `y[k] = g[k+1] - g[k]`.
+    rho_history: array containing `rho[k] = 1. / vdot(s[k], y[k])`.
+    gamma: scalar to use for the initial inverse Hessian approximation,
+      i.e., `gamma * I`.
+
+  Reference:
+    Jorge Nocedal and Stephen Wright.
+    Numerical Optimization, second edition.
+    Algorithm 7.4 (page 178).
+  """
+  fun = partial(inv_hessian_product_leaf, rho_history=rho_history, gamma=gamma)
+  return tree_map(fun, pytree, s_history, y_history)
+
+
+def compute_gamma(s_history: Any, y_history: Any):
+  # Let gamma = vdot(y_history[-1], s_history[-1]) / sqnorm(y_history[-1]).
+  # The initial inverse Hessian approximation can be set to gamma * I.
+  # See Numerical Optimization, second edition, equation (7.20).
+  # Note that unlike BFGS, the initialization can change on every iteration.
+
+  fun = lambda s_history, y_history: tree_vdot(y_history[-1], s_history[-1])
+  num = tree_sum(tree_map(fun, s_history, y_history))
+
+  fun = lambda y_history: tree_vdot(y_history[-1], y_history[-1])
+  denom = tree_sum(tree_map(fun, y_history))
+
+  return jnp.where(denom > 0, num / denom, 1.0)
+
+
+def init_history(pytree, history_size):
+  fun = lambda leaf: jnp.zeros((history_size,) + leaf.shape)
+  return tree_map(fun, pytree)
+
+
+def _update_history(history_array, new_value):
+  """Shift past elements to the left and add new elements at the end."""
+  # TODO: to avoid memory copies, it would be more efficient to treat history as
+  # a rolling buffer, and only set a single vector.
+  return jnp.roll(history_array, -1, axis=0).at[-1].set(new_value)
+
+
+def update_history(history_pytree, new_pytree):
+  return tree_map(_update_history, history_pytree, new_pytree)
+
+
+class LbfgsState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  stepsize: float
+  error: float
+  s_history: Any
+  y_history: Any
+  rho_history: jnp.ndarray
+  aux: Optional[Any] = None
+
+
+@dataclass(eq=False)
+class LBFGS(base.IterativeSolver):
+  """LBFGS solver.
+
+  Attributes:
+    fun: a smooth function of the form ``fun(x, *args, **kwargs)``.
+    has_aux: whether function fun outputs one (False) or more values (True).
+      When True it will be assumed by default that fun(...)[0] is the objective.
+
+    maxiter: maximum number of proximal gradient descent iterations.
```

Comment:
spacing is strange here. why are some parameters clustered (sometime there's a blank line between attributes but not always)?

Reply:
I grouped options by "category" because I think very long lists of attributes without any line spacing are very difficult to read.

Reply:
sounds good